--- ../../../source/nova-19.0.1/nova/virt/libvirt/vif.py	2019-06-06 06:47:33.000000000 +0000
+++ ../../nova-19.0.1/nova/virt/libvirt/vif.py	2019-08-29 07:04:52.604855296 +0000
@@ -222,11 +222,12 @@
         if img_props.get('hw_vif_multiqueue_enabled'):
             driver = 'vhost'
             max_tap_queues = self._get_max_tap_queues()
-            if max_tap_queues:
-                vhost_queues = (max_tap_queues if flavor.vcpus > max_tap_queues
-                    else flavor.vcpus)
-            else:
-                vhost_queues = flavor.vcpus
+            # TODO(hyphon): e.g. [0,8)->1; [8,16)->2; [16,32)->4; [32,64)->8;
+            import math
+            vhost_queues = (1 if flavor.vcpu < 8
+                else pow(2, int(math.log(flavor.vcpu, 2)) - 2))
+            if max_tap_queues and max_tap_queues < vhost_queues:
+                vhost_queues = max_tap_queues
 
         return (driver, vhost_queues)
 
